.github/workflows/python_tests.yml

```diff
@@ -35,6 +35,12 @@ jobs:
       - uses: actions/setup-python@v2
         with:
           python-version: ${{ matrix.python-version }}
+      - uses: actions/cache@v2
+        with:
+          path: ~/infogami_python_venv
+          key: ${{ runner.os }}-venv-${{ hashFiles('*requirements.txt') }}
+      - run: python3 -m venv ~/infogami_python_venv
+      - run: source ~/infogami_python_venv/bin/activate
       - run: pip install black codespell flake8 isort mypy pytest pyupgrade
```

Comment:
You need to source venv/bin/activate for every step. Each step is completely independent and the environment of a step is not passed to the next step. 

```
    - run: |
        source ~/infogami_python_venv/bin/activate
        pip install black codespell flake8 isort mypy pytest pyupgrade
    - run: |
        source ~/infogami_python_venv/bin/activate
        black --check . || true
```

Reply:
It may also be a good idea to add a dev-requirements.txt file.

Reply:
Yes, that and I am also thinking of grouping all the `run` steps into a bash script in the `scripts/` directory to keep the workflow file simple.